File: Frontend/models/emergency_alert.dart
class EmergencyAlert {
  final String id;
  final String studentId;
  final double latitude;
  final double longitude;
  final String status;
  final String? assignedProctor;
  final String? acknowledgedByUserId;
  final String? acknowledgedByName;
  final String? acknowledgedAt;
  final double? distanceInKm;
  final Map<String, dynamic>? responderLocation;
  final String createdAt;
  final Map<String, dynamic>? student;
  final Map<String, dynamic>? acknowledgedBy;

  EmergencyAlert({
    required this.id,
    required this.studentId,
    required this.latitude,
    required this.longitude,
    required this.status,
    this.assignedProctor,
    this.acknowledgedByUserId,
    this.acknowledgedByName,
    this.acknowledgedAt,
    this.distanceInKm,
    this.responderLocation,
    required this.createdAt,
    this.student,
    this.acknowledgedBy,
  });

  factory EmergencyAlert.fromJson(Map<String, dynamic> json) {
    return EmergencyAlert(
      id: json['id'],
      studentId: json['student_id'],
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      status: json['status'] ?? 'active',
      assignedProctor: json['assigned_proctor'],
      acknowledgedByUserId: json['acknowledged_by_user_id']?.toString(),
      acknowledgedByName: json['acknowledged_by_name']?.toString(),
      acknowledgedAt: json['acknowledged_at']?.toString(),
      distanceInKm: json['distance_in_km'] == null
          ? null
          : double.tryParse(json['distance_in_km'].toString()),
      responderLocation: json['responder_location'] is Map<String, dynamic>
          ? json['responder_location'] as Map<String, dynamic>
          : null,
      createdAt: json['created_at'] ?? '',
      student: json['student'],
      acknowledgedBy: json['acknowledgedBy'] is Map<String, dynamic>
          ? json['acknowledgedBy'] as Map<String, dynamic>
          : null,
    );
  }
}

File: backend/src/models/EmergencyAlert.js
const { DataTypes } = require('sequelize');
const sequelize = require('../config/database');

const EmergencyAlert = sequelize.define('EmergencyAlert', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  student_id: {
    type: DataTypes.UUID,
    allowNull: false,
    references: { model: 'users', key: 'id' },
  },
  latitude: {
    type: DataTypes.DOUBLE,
    allowNull: false,
  },
  longitude: {
    type: DataTypes.DOUBLE,
    allowNull: false,
  },
  assigned_proctor: {
    type: DataTypes.UUID,
    allowNull: true,
    references: { model: 'users', key: 'id' },
  },
  acknowledged_by_user_id: {
    type: DataTypes.UUID,
    allowNull: true,
    references: { model: 'users', key: 'id' },
  },
  acknowledged_by_name: {
    type: DataTypes.STRING(120),
    allowNull: true,
  },
  acknowledged_at: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  responder_location: {
    type: DataTypes.JSONB,
    allowNull: true,
  },
  distance_in_km: {
    type: DataTypes.DECIMAL(8, 3),
    allowNull: true,
  },
  status: {
    type: DataTypes.ENUM('active', 'acknowledged', 'resolved'),
    defaultValue: 'active',
  },
}, {
  tableName: 'emergency_alerts',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: false,
});

module.exports = EmergencyAlert;
